### PR TITLE
Add drop_example flag to the RunInference and Model Handler

### PR DESCRIPTION
**Pull request by AnandInguva**
Fixes: https://github.com/apache/beam/issues/21444

 Added a flag `drop_example`. This can be passed to RunInference API to drop the example from the PredictionResult.

```
if drop_example is True, prediction_result.example would be None
```

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by AnandInguva** [comment]
Run Python 3.9 PostCommit

**Comment by AnandInguva** [comment]
One more way to do this is to pass the flag through the `ModelHandler`. The reason why I added it in the RunInference because RunInference emits PredictionResult and I thought it would be the appropriate place to control how the output[prediction result] should look.

**Comment by damccorm** [comment]
> One more way to do this is to pass the flag through the `ModelHandler`. The reason why I added it in the RunInference because RunInference emits PredictionResult and I thought it would be the appropriate place to control how the output[prediction result] should look.

I agree

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:105]: Rather than returning from inside this nested loop, could we just set `predictions = [dict(zip(predictions.keys(), v)) for v in zip(*predictions.values())]` and let the if/else handle it below? That way we don't need to duplicate logic

File: sdks/python/apache_beam/ml/inference/base.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base.py:85]: Good call moving this out to the base class.

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/base_test.py:82]: Isn't this the behavior we're testing? Why are we baking it into our mock?

File: sdks/python/apache_beam/ml/inference/sklearn_inference.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/sklearn_inference.py:171]: Don't we need to pass this into _convert_to_result? Same question exists elsewhere too

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/sklearn_inference.py:171]: Done

File: sdks/python/apache_beam/ml/inference/base.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base.py]: Done. PTAL

File: sdks/python/apache_beam/ml/inference/base.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/base.py:122]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/base.py:424]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/base.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/base.py:317]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/pytorch_inference.py:265]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/pytorch_inference.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/pytorch_inference.py:156]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/sklearn_inference.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/sklearn_inference.py:110]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/sklearn_inference.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/sklearn_inference.py:181]:
```suggestion
        drop_example: Boolean flag indicating whether or not to drop the example from PredictionResult
```

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/base_test.py:289]: Shouldn't this have `drop_example=True`?

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (damccorm) [reply sdks/python/apache_beam/ml/inference/base_test.py:82]: I think we should be calling `_convert_to_result` here and yielding that

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base_test.py]: Done

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (yeandy) [inline sdks/python/apache_beam/ml/inference/base_test.py:59]: The name of this class was initially confusing to me. I think I see what you mean since `run_inference` returns `PredictionResult`, but I'm wondering if there's a better way to word this. Maybe `FakeModelHandlerReturnsPredictionResult`?

**Review by yeandy — COMMENTED** [review]
If this gets merged before TensorRT, we'll need to remember to rebase that.

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base_test.py:289]: We are passing `drop_example` via `base.RunInference`, which passes it to the `ModelHandler`.

More explanation: https://github.com/apache/beam/pull/23266#issuecomment-1248854707

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/base_test.py:59]: Agreed

**Comment by AnandInguva** [comment]
PTAL @yeandy @damccorm

File: sdks/python/apache_beam/ml/inference/base_test.py
Comment (yeandy) [reply sdks/python/apache_beam/ml/inference/base_test.py]: Thanks! I missed seeing the `drop=True` in the test example  😄

**Review by yeandy — APPROVED** [review]
LGTM

**Comment by damccorm** [comment]
> If this gets merged before TensorRT, we'll need to remember to rebase that.

I'd vote we let that one merge first since it is the higher complexity change (and less easy for us to control). Its currently just blocked on Python PreCommit flakes I think

**Review by damccorm — COMMENTED** [review]
I know it came in after this PR, but could you please extend these changes to [tensorrt_inference.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/ml/inference/tensorrt_inference.py) as well?

**Comment by AnandInguva** [comment]
Run Python 3.8 PostCommit

**Comment by AnandInguva** [comment]
stop reviewer notifications

File: sdks/python/apache_beam/ml/inference/tensorrt_inference.py
Comment (damccorm) [inline sdks/python/apache_beam/ml/inference/tensorrt_inference.py:277]: Can we use `_convert_to_result` here too? I don't think there's a reason these predictions can't be dictionaries, so this will handle that case automatically

File: sdks/python/apache_beam/ml/inference/tensorrt_inference.py
Comment (AnandInguva) [inline sdks/python/apache_beam/ml/inference/tensorrt_inference.py:275]: PTA:L @damccorm I think this is how we use `_convert_to_result`

Comment (AnandInguva) [reply sdks/python/apache_beam/ml/inference/tensorrt_inference.py:275]: cc : @yeandy

**Comment by damccorm** [comment]
That looks right to me, lets run the PostCommit as well though

**Comment by damccorm** [comment]
Run Python 3.8 PostCommit

**Comment by damccorm** [comment]
@AnandInguva looks like there's a linting violation. If that is fixed and we have a successful postcommit run I think this should be good to merge.

**Comment by AnandInguva** [comment]
https://ci-beam.apache.org/job/beam_PostCommit_Python38_PR/603/console passes

**Comment by damccorm** [comment]
Run Python PreCommit

**Review by damccorm — APPROVED** [review]
LGTM - will merge once we get past this flaky precommit

**Pull request merged**
